website_compassion/controllers/my_account.py

```diff
@@ -385,6 +386,90 @@ def my_donations(self, form_id=None, **kw):
             )
         return self._form_redirect(result, full_page=True)
 
+    @route("/my/new/donation", type="http", auth="user", website=True)
+    def make_new_donation(self, **kwargs):
+        """
+        The route to display a form with the possibility to make a new
+        donation to a compassion fund
+        :return: a redirection to a webpage
+        """
+        available_funds = request.env["product.product"].search([("website_published", "=", True)])
+
+        values = self._prepare_portal_layout_values()
+
+        values.update({
+            "funds": available_funds
+        })
+
+        result = request.render(
+            "website_compassion.my_new_donation_template", values
+        )
+
+        return self._form_redirect(result, full_page=True)
+
+    @route("/my/new/donation/fund/<int:selected_fund_id>", type="http", auth="user", website=True)
+    def make_new_donation_to_fund(self, selected_fund_id=None, **kwargs):
+        partner = request.env.user.partner_id
+        available_funds = request.env["product.product"].search([("website_published", "=", True)])
+
+        if not selected_fund_id or int(selected_fund_id) not in available_funds.mapped("id"):
+            return request.redirect("/my/new/donation")
+
+        selected_fund_id = available_funds.browse(int(selected_fund_id))
+
+        values = kwargs.copy()
+        # This allows the translation to still work on the page
+        values.pop("edit_translations", False)
+        values.update({
+            "form_model_key": "cms.form.fund.donation",
+            "selected_fund": selected_fund_id,
+        })
+
+        donation_form = self.get_form("res.partner", partner.id, **values)
+        donation_form.form_process()
+
+        values.update({
+            "form": donation_form,
+        })
+
+        # If the form is valid, redirect to payment
+        if donation_form.form_success:
+            return werkzeug.utils.redirect(donation_form.form_next_url(), code=303)
+
+        values.update(self._prepare_portal_layout_values())
+
+        result = request.render(
+            "website_compassion.new_donation_to_fund_template", values
+        )
+
+        return self._form_redirect(result, full_page=True)
+
+    @route("/my/new/donation/validate/<int:invoice_id>", type="http", auth="user", website=True)
+    def new_donation_validation(self, invoice_id=None, **kwargs):
+
+        payment = kwargs.get("payment")
+        values = kwargs.copy()
+
+        values.update(self._prepare_portal_layout_values())
+
+        try:
+            invoice = request.env["account.invoice"].sudo().browse(int(invoice_id))
+            invoice.exists().ensure_one()
+            transaction = invoice.get_portal_last_transaction()
+        except ValueError:
+            transaction = request.env["payment.transaction"]
```

Comment:
Maybe it would be better, if possible, to do an 'if statement' here instead of a try, except. Or raise an error if it is an unexpected behaviour.

Reply:
This code can also be found in other module. If getting the last transaction cause some issue it will raise a Value Error. To avoid further problem in the function logic we set transaction to an empty one. This will not be an issue because during the next step we check that its state is equal to "done" or we redirect to the error page